File: app/javascript/channels/spreadsheet_channel.js
import consumer from "./consumer"

// Store active spreadsheet subscription
let spreadsheetSubscription = null;

export function subscribeToSpreadsheet(spreadsheetId, sheetId) {
  if (spreadsheetSubscription) {
    spreadsheetSubscription.unsubscribe();
  }

  spreadsheetSubscription = consumer.subscriptions.create(
    {
      channel: "SpreadsheetChannel",
      spreadsheet_id: spreadsheetId,
      sheet_id: sheetId
    },
    {
      connected() {
        console.log("Connected to spreadsheet channel");
      },

      disconnected() {
        console.log("Disconnected from spreadsheet channel");
      },

      received(data) {
        console.log("Received data:", data);

        switch(data.type) {
          case 'cell_update':
            handleCellUpdate(data);
            break;
          case 'row_added':
            handleRowAdded(data);
            break;
          case 'column_added':
            handleColumnAdded(data);
            break;
          case 'cursor_position':
            handleCursorPosition(data);
            break;
        }
      },

      sendCellUpdate(rowId, columnKey, value, formula) {
        this.perform('receive', {
          action: 'cell_update',
          row_id: rowId,
          column_key: columnKey,
          value: value,
          formula: formula
        });
      },

      sendCursorPosition(rowId, columnKey) {
        this.perform('receive', {
          action: 'cursor_position',
          row_id: rowId,
          column_key: columnKey
        });
      }
    }
  );

  return spreadsheetSubscription;
}

function handleCellUpdate(data) {
  const cell = document.querySelector(
    `[data-row="${data.row_id}"][data-column="${data.column_key}"]`
  );

  if (cell && document.activeElement !== cell) {
    cell.textContent = data.value || data.formula;
    cell.dataset.value = data.value;
    cell.dataset.formula = data.formula;

    // Visual feedback for remote update
    cell.classList.add('remote-update');
    setTimeout(() => cell.classList.remove('remote-update'), 1000);
  }
}

function handleRowAdded(data) {
  // Reload page or dynamically add row
  console.log("Row added by another user:", data);
}

function handleColumnAdded(data) {
  // Reload page or dynamically add column
  console.log("Column added by another user:", data);
}

function handleCursorPosition(data) {
  // Show other users' cursors
  const cell = document.querySelector(
    `[data-row="${data.row_id}"][data-column="${data.column_key}"]`
  );

  if (cell) {
    showRemoteCursor(cell, data.user_name);
  }
}

function showRemoteCursor(cell, userName) {
  // Remove existing cursor indicator
  const existingCursor = cell.querySelector('.remote-cursor');
  if (existingCursor) {
    existingCursor.remove();
  }

  // Add cursor indicator
  const cursor = document.createElement('div');
  cursor.className = 'remote-cursor';
  cursor.textContent = userName;
  cell.appendChild(cursor);

  // Remove after 3 seconds
  setTimeout(() => cursor.remove(), 3000);
}

File: app/channels/spreadsheet_channel.rb
class SpreadsheetChannel < ApplicationCable::Channel
  def subscribed
    spreadsheet = Spreadsheet.find(params[:spreadsheet_id])
    sheet = spreadsheet.sheets.find(params[:sheet_id])

    # Subscribe to this specific sheet's updates
    stream_for sheet
  end

  def unsubscribed
    # Any cleanup needed when channel is closed
  end

  def receive(data)
    # Handle incoming messages from clients
    case data['action']
    when 'cell_update'
      broadcast_cell_update(data)
    when 'row_added'
      broadcast_row_added(data)
    when 'column_added'
      broadcast_column_added(data)
    when 'cursor_position'
      broadcast_cursor_position(data)
    end
  end

  private

  def broadcast_cell_update(data)
    sheet = Sheet.find(params[:sheet_id])
    SpreadsheetChannel.broadcast_to(
      sheet,
      {
        type: 'cell_update',
        row_id: data['row_id'],
        column_key: data['column_key'],
        value: data['value'],
        formula: data['formula'],
        user_id: current_user&.id || 'anonymous'
      }
    )
  end

  def broadcast_row_added(data)
    sheet = Sheet.find(params[:sheet_id])
    SpreadsheetChannel.broadcast_to(
      sheet,
      {
        type: 'row_added',
        row: data['row'],
        user_id: current_user&.id || 'anonymous'
      }
    )
  end

  def broadcast_column_added(data)
    sheet = Sheet.find(params[:sheet_id])
    SpreadsheetChannel.broadcast_to(
      sheet,
      {
        type: 'column_added',
        column: data['column'],
        user_id: current_user&.id || 'anonymous'
      }
    )
  end

  def broadcast_cursor_position(data)
    sheet = Sheet.find(params[:sheet_id])
    SpreadsheetChannel.broadcast_to(
      sheet,
      {
        type: 'cursor_position',
        row_id: data['row_id'],
        column_key: data['column_key'],
        user_id: current_user&.id || 'anonymous',
        user_name: current_user&.email || 'Anonymous'
      }
    )
  end
end
